src/content/resources/javascript/promises/async-await.md

```diff
@@ -0,0 +1,137 @@
+---
+authors:
+  - "ddivad#1337"
+created_at: "2019/10/01"
+title: Async Await
+recommended_reading:
+  - javascript/promises
+---
+
+## Async Await
+
+Async / Await is another way to handle the calling of asynchronous code and is built on top of Promises. It is a nice and clean way to handle asynchronous code, without the need to `.then()` functions to the promise.
+
+It is made up of 2 keywords as the name suggests: `async` and `await`. These need to be used together and can for this method to work.
+
+## Async
+
+The `async` keyword is used to show that the function is going to return a promise. Any return values from the function will be converted to a promise automatically, if they are not already.
+
+```js
+async function foo() {
+  return "foo";
+}
+```
+
+This basic example will return a promise with the value of `foo`.
+
+## Await
+
+The `await` keyword is used to wait until a promise has executed and fetches the result. In order to use the `await` keyword, you **need** to be inside an `async` function. 
+
+```js
+let data = await foo();
+console.log(foo); // "foo"
+```
+
+This code will call the `foo()` function from above and will wait on that line until the promise returned from the `async` function has returned.
+
+Another example to demonstrate this is if we extended the `foo()` function from above to add a 5 second delay.
+
+```js
+async function foo() {
+  let promise = new Promise((resolve, reject) => {
+    setTimeout(() => resolve("foo"), 5000)
+  });
+
+  let data = await promise; // Execution waits here until promise resolves
+  console.log(data); // "foo"
+}
+```
+
+## Error Handling
+
+As with Promises, if the Promise returns with an error, it will affect the `await` call that triggered it. When using Promises without Async / Await, the `.catch()` syntax is used, and this also exists with Aysnc / Await using the `try...catch` syntax. This can be demonstrated by making our `foo()` function throw an error.
+
+```js
+async function foo() {
+  throw new Error("An Error Occurred");
+}
+```
+
+This would be handled using Aysnc / Await like so:
+
+```js
+async function getData() {
+  try {
+    let data = await foo();
+    console.log(data);
+  } catch (err) {
+    console.error(err); // err: An Error Occurred
+  }
+}
+```
+
+## Beginner Mistakes
```

Comment:
This section should absolutely be mentioning the mistake of wrapping promises inside `new Promise`, that's something I've seen too many times.

Reply:
Although we might just want to edit the original promise article for that instead, I guess I could see that going either way.

Reply:
Good point - I guess adding in to the original promise article might be better here, as I have that article listed in the recommended reading section, and can add that one there explicitly. 

People should read that one first, so the way I was approaching this section here was like an addition to the common mistakes from the original promise article, rather than adding them in both places.